### PR TITLE
[libc][C23][math] Implement cospif function correctly rounded for all rounding modes

### PR DESCRIPTION
**Pull request by HendrikHuebner**
I also fixed a comment in sinpif.cpp in the first commit. Should this be included in this PR?

All tests were passed, including the exhaustive test.

CC: @lntue

### CODE REVIEW

**Comment by HendrikHuebner** [comment]
The Core-Math tests have passed

File: libc/test/src/math/smoke/cospif_test.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+  EXPECT_FP_EQ(1.0f, LIBC_NAMESPACE::cospif(-0.0f));
+  EXPECT_MATH_ERRNO(0);
+
+  EXPECT_FP_EQ(aNaN, LIBC_NAMESPACE::cospif(inf));
+  EXPECT_MATH_ERRNO(EDOM);
+
+  EXPECT_FP_EQ(aNaN, LIBC_NAMESPACE::cospif(neg_inf));
+  EXPECT_MATH_ERRNO(EDOM);
+}
```
Comment (lntue) [inline libc/test/src/math/smoke/cospif_test.cpp:34]: need extra empty line at the end of the file.

**Comment by lntue** [comment]
@zimmermann6

**Comment by zimmermann6** [comment]
I get a failure for rounding down:
```
Running exhaustive check in --rndn mode...
all ok
Running exhaustive check in --rndz mode...
all ok
Running exhaustive check in --rndu mode...
all ok
Running exhaustive check in --rndd mode...
FAIL x=0x1.7408p+12 ref=0x0p+0 y=-0x0p+0
```

**Comment by HendrikHuebner** [comment]
hm, I'm not sure why those tests are failing. They pass when I run them locally.

**Comment by HendrikHuebner** [comment]
Yeah I'm not getting any failures running` ./check.sh --exhaustive cospif`. Are you on the latest commit @zimmermann6 ?

**Comment by zimmermann6** [comment]
I got a similar error with revision 94012aa:
```
Running exhaustive check in --rndd mode...
FAIL x=0x1.6c0008p+20 ref=0x0p+0 y=-0x0p+0
```
This is on a AMD EPYC 7282. I can send you my libllvmlibc.a by mail if you want.

**Comment by HendrikHuebner** [comment]
I can test it on an AMD EPYC 7713P later. Do you know what the buildbot runs on? It keeps failing as well which I also can't reproduce

**Comment by zimmermann6** [comment]
no I don't know for the buildbot, sorry

**Comment by lntue** [comment]
@HendrikHuebner you can add the following test cases to smoke tests:
```
  EXPECT_FP_EQ_ALL_ROUNDING(zero, LIBC_NAMESPACE::cospif(0.5f));
  EXPECT_FP_EQ_ALL_ROUNDING(neg_zero, LIBC_NAMESPACE::cospif(-0.5f));
```

File: libc/src/math/generic/cospif.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  // Combine the results with the sine of sum formula:
+  //   cos(pi * x) = cos((k + y)*pi/32)
+  //          = cos(y*pi/32) * cos(k*pi/32) - sin(y*pi/32) * sin(k*pi/32)
+  //          = (cosm1_y + 1) * cos_k - sin_y * sin_k
+  //          = (cosm1_y * cos_k + cos_k) - sin_y * sin_k
+  double sin_k, cos_k, sin_y, cosm1_y;
+
+  sincospif_eval(xd, sin_k, cos_k, sin_y, cosm1_y);
+
```
Comment (lntue) [inline libc/src/math/generic/cospif.cpp:86]: Adding a similar test to your `sinpif` should fix the signed zero issue:
```
  if (LIBC_UNLIKELY(sin_y == 0 && cos_k == 0))
    return FPBits::zero(xbits.sign()).get_val();
```

**Comment by HendrikHuebner** [comment]
Okay, I fixed the signed zero case. The buildbot passes now as well. My mpfr implementation for older mpfr version did not handle integer inputs correctly.

**Review by lntue — APPROVED** [review]

**Comment by zimmermann6** [comment]
I still get errors for n+1/2 when n is negative. IEEE 754-2019 says that cos(n+1/2) is always +0, whether n is positive or negative, so that cos(-x) = cos(x) always holds.
```
Running exhaustive check in --rndn mode...
FAIL x=-0x1.7408p+12 ref=0x0p+0 y=-0x0p+0
Running exhaustive check in --rndz mode...
FAIL x=-0x1.70008p+16 ref=0x0p+0 y=-0x0p+0
Running exhaustive check in --rndu mode...
FAIL x=-0x1.7408p+12 ref=0x0p+0 y=-0x0p+0
Running exhaustive check in --rndd mode...
FAIL x=-0x1.8p+0 ref=0x0p+0 y=-0x0p+0
```

**Comment by zimmermann6** [comment]
all exhaustive tests are ok now, thanks!

**Comment by HendrikHuebner** [comment]
Have you tested with the commit from my new PR?  #98037

**Comment by zimmermann6** [comment]
yes I tested with #98037, sorry my comment was in the wrong place

**Pull request merged**